src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.cs

```diff
@@ -0,0 +1,135 @@
+            RuleIdTwoParameterId,
+            s_localizableTitle2,
+            s_localizableMessageDefault2,
+            DiagnosticCategory.Usage,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription2,
+            helpLinkUri: "", // TODO (might require an addition to docs repository?
+            customTags: WellKnownDiagnosticTags.AnalyzerException);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(RuleReplaceOneParameter, RuleReplaceTwoParameter);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSyntaxNodeAction(this.Action, SyntaxKind.InvocationExpression);
+        }
+
+        private void Action(SyntaxNodeAnalysisContext context)
+        {
+            if (context.Node is InvocationExpressionSyntax invocation)
+            {
+                var invokedSymbol = context.SemanticModel.GetSymbolInfo(invocation);
+                var invokedClass = invokedSymbol.Symbol.ContainingSymbol;
+                var invokedMethod = invokedSymbol.Symbol.Name;
+                var namedType = invokedClass as INamedTypeSymbol;
+
+                if (namedType.ContainingNamespace.Name == "Text" 
+                    && namedType.ContainingNamespace.ContainingNamespace.Name == "System"
+                    && namedType.ContainingNamespace.ContainingNamespace.ContainingNamespace.IsGlobalNamespace
+                    && invokedClass.Name == "StringBuilder" 
+                    && invokedMethod == "Append"
+                    && invokedSymbol.Symbol is IMethodSymbol methodSymbol)
+                {
+                    var parameters = methodSymbol.Parameters;
+
+                    if (parameters.Length == 1 && parameters[0].Type.Name == "String")
+                    {
+                        ArgumentSyntax argument = invocation.ArgumentList.Arguments.First();
+
+                        // todo: what if there's ref? argument.RefOrOutKeyword
+                        if (argument.Expression is InvocationExpressionSyntax invocationExpression)
+                        {
+                            var innerInvokedSymbol = context.SemanticModel.GetSymbolInfo(invocationExpression);
+                            if (innerInvokedSymbol.Symbol is IMethodSymbol parameterMethod
+                                && parameterMethod.ContainingSymbol.Name == "String"
+                                && parameterMethod.ContainingSymbol.ContainingNamespace.Name == "System"
+                                && parameterMethod.ContainingSymbol.ContainingNamespace.ContainingNamespace.IsGlobalNamespace
+                                && parameterMethod.Name == "Substring")
```

Comment:
can just check `.SpecialType == SpecialType.String` on appropriate parts of this.